### ISSUE TITLE
Implement `org.freedesktop.impl.portal.Secret`

### ISSUE THREAD

**Issue by q234rty** [keepassxreboot/keepassxc#8858]
## Summary

See the documentation at https://flatpak.github.io/xdg-desktop-portal/#gdbus-org.freedesktop.impl.portal.Secret

This is supported by libsecret and is supposed to be used instead of `org.freedesktop.secrets` for sandboxed applications.

## Context
Any flatpak app that uses libsecret and does not require access to `org.freedesktop.secrets` needs this.

**Comment by hpfr** [comment]
Looks like the docs moved to https://flatpak.github.io/xdg-desktop-portal/docs/doc-org.freedesktop.impl.portal.Secret.html

It looks like the interface is quite simple, just store a “master secret” the client will use to encrypt arbitrary data, generating the secret on first request. It is a single D-Bus method, looks like apps pass a file descriptor expecting the server to write the secret there.

There is an implementation of the server side at https://github.com/bilelmoussaoui/oo7/tree/main/portal. This server implementation acts as an `org.freedesktop.secrets` client itself, storing and retrieving the generated secret in the default collection. So if you have KeePassXC running with the secret service enabled, I believe you can install and run `oo7-portal` and it will communicate with `org.freedesktop.portal.Secret` clients to store their master secrets in KeePassXC.

Note the portal needs to be registered with `xdg-desktop-portal`, which the oo7 build system and your distro’s package definition should handle. I tried this with NixOS via `xdg.portal.extraPortals = [ pkgs.oo7 ];` but it doesn’t seem to activate via D-Bus when launching the Fractal flatpak and it still doesn’t seem to work when running oo7-portal manually. nixpkgs currently packages the latest oo7 tag (0.3.3), and fairly substantial changes since then are on the main branch.

Setting aside the D-Bus boilerplate, the server is basically [a single function](https://github.com/bilelmoussaoui/oo7/blob/37198e79b0503c60cf19d1401ce7f685efd3e76a/portal/src/main.rs#L47). Presumably similar functionality could be implemented directly within KeePassXC without too much trouble.

@bilelmoussaoui, I see your :+1: here. Please let me know if I misunderstood anything in this comment. I guess you think it makes sense for KeePassXC to ship a built-in secret portal backend given it already includes a much larger `org.freedesktop.secrets` server implementation? Thank you! PS: could we get a new oo7 tag?

**Comment by bilelmoussaoui** [comment]
> @bilelmoussaoui, I see your 👍 here. Please let me know if I misunderstood anything in this comment. I guess you think it makes sense for KeePassXC to ship a built-in secret portal backend given it already includes a much larger org.freedesktop.secrets server implementation? Thank you! PS: could we get a new oo7 tag?

Yes, what you described above is correct. The portal is just a way for apps to store a key on the host keyring used to encrypt a file that is usually used to store any sensitive information without giving the app access to the whole host keyring.

For oo7 tag, it will happen in the future but I am waiting first to see if there would be any breaking changes that might be needed for the to be finished org.freedesktop.secrets server implementation.

**Comment by zroug** [comment]
Thanks for pointing me to oo7. I tried the portal service from the current main branch and it works just fine with KeePassXC.

As to whether support for this portal should be built into KeePassXC: I think this portal is very important, because giving all applications that need to store a secret access to all other applications' secrets is not really acceptable. So for the end user the portal should 'just work'. If it is kept separate from KeePassXC, there should be some kind of default dependency on a portal provider at install time, which may not be ideal.

I guess ideally the desktop environments would just allow you to swap out the `org.freedesktop.secrets` provider and provide everything else (including any services native to the specific desktop environment) based on that. But that's probably not going to happen.

**Comment by MuntasirSZN** [comment]
@bilelmoussaoui Sorry for ping but any update?
